Publish Docker image workflow fails on push events due to invalid GHA_CR_TOKEN

## Summary

The "Publish Docker image" workflow fails on all `push` events because the `GHA_CR_TOKEN` secret is invalid or expired. Pull request events succeed because they skip the ghcr.io push entirely.

## Evidence

### Pattern analysis of recent runs:

| Event Type | Result |
|------------|--------|
| `pull_request` | ✅ Always succeeds |
| `push` | ❌ Always fails |

### Error from logs:

```
Logging into ghcr.io...
##[error]Error response from daemon: Get "https://ghcr.io/v2/": denied: denied
```

### Workflow logic (lines 72-77):

```python
if gh_event == 'push' and has_cr_token == 'true':
    print("Adding GitHub Container Repository (ghcr.io)")
    i.append("ghcr.io/{}/image".format(gh_repo))
else:
    print("Skipping GitHub Container Repository (ghcr.io)")
```

- `HAS_GHA_CR_TOKEN` is `true` (secret exists)
- But authentication fails → token is invalid/expired or lacks `write:packages` permission

## Root Cause

The workflow uses a custom Personal Access Token (`GHA_CR_TOKEN`) which is an older authentication pattern. The token appears to be invalid or expired.

## Suggested Fix

Migrate to the modern `GITHUB_TOKEN` approach:

1. Add permissions block to the workflow:
   ```yaml
   permissions:
     contents: read
     packages: write
   ```

2. Change the login step to use `GITHUB_TOKEN`:
   ```yaml
   - name: Login to GHCR
     uses: docker/login-action@v1
     with:
       registry: ghcr.io
       username: ${{ github.actor }}
       password: ${{ secrets.GITHUB_TOKEN }}
   ```

3. Simplify the "Push To" logic to always include ghcr.io for push events (no need to check for custom token)

## Benefits of GITHUB_TOKEN approach

- No manual token management required
- Token is automatically rotated
- Proper permissions scoped per-workflow
- Works for both repository owner and forks (with appropriate permissions)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

--- a/.github/workflows/publish-docker-image.yml
+++ b/.github/workflows/publish-docker-image.yml
@@ -26,6 +26,10 @@ jobs:
     name: Push Docker image to GitHub Packages
     runs-on: ubuntu-latest
 
+    permissions:
+      contents: read
+      packages: write
+
     # Run a local registry
     services:
       registry:
@@ -51,8 +55,6 @@ jobs:
     - name: Push To
       id: push_to
       shell: python
-      env:
-        HAS_GHA_CR_TOKEN: ${{ secrets.GHA_CR_TOKEN != '' }}
       run: |
         import os
 
@@ -63,22 +65,25 @@ jobs:
 
         gh_repo = g('GITHUB_REPOSITORY')
         gh_event = g('GITHUB_EVENT_NAME')
-        has_cr_token = g('HAS_GHA_CR_TOKEN')
 
         i = []
 
         print("Adding local service.")
         i.append("localhost:5000/"+gh_repo)
 
-        if gh_event == 'push' and has_cr_token == 'true':
+        # Use GITHUB_TOKEN for authentication (always available)
+        if gh_event == 'push':
             print("Adding GitHub Container Repository (ghcr.io)")
             i.append("ghcr.io/{}/image".format(gh_repo))
         else:
-            print("Skipping GitHub Container Repository (ghcr.io)")
+            print("Skipping GitHub Container Repository (ghcr.io) for non-push events")
 
         l = ",".join(i)
         print("Final locations:", repr(l))
-        print("::set-output name=images::{}".format(l))
+
+        # Use environment file instead of deprecated set-output
+        with open(os.environ['GITHUB_OUTPUT'], 'a') as f:
+            f.write("images={}\n".format(l))
 
     - name: Docker meta
       id: docker_meta
@@ -97,11 +102,11 @@ jobs:
 
     - name: Login to GHCR
       if: ${{ contains(steps.push_to.outputs.images, 'ghcr.io') }}
-      uses: docker/login-action@v1
+      uses: docker/login-action@v3
       with:
-        username: ${{ github.actor }}
-        password: ${{ secrets.GHA_CR_TOKEN }}
         registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
 
 #    - name: Login to local registry
 #      uses: docker/login-action@v1

--- a/.gitignore
+++ b/.gitignore
@@ -140,3 +140,6 @@ cython_debug/
 # Generated version file
 actions_includes/version.py
 docker/*.tar.gz
+
+# Git worktrees
+.worktrees/

